<?xml version="1.0" encoding="utf-8"?>
<launch>
  <arg name="INPUT_CLOUD" default="/rs_d435/depth_registered/points" />
  <rosparam file="$(find aerial_robot_perception)/config/HorizontalObjectDetectionConfig.yaml" command="load"/>
  <node name="horizontal_object_detector_manager"
        pkg="nodelet" type="nodelet"
        args="manager"/>

  <node name="passthrough_filter_for_object"
        pkg="nodelet" type="nodelet"
        args="load pcl/PassThrough /horizontal_object_detector_manager">
    <remap from="~input" to="$(arg INPUT_CLOUD)" />
  </node>

  <node pkg="nodelet" type="nodelet" name="hsi_filter_red" respawn="true"
        args="load jsk_pcl/HSIColorFilter horizontal_object_detector_manager" >
    <remap from="~input" to="passthrough_filter_for_object/output" />
  </node>

  <node name="passthrough_filter_for_ground"
        pkg="nodelet" type="nodelet"
        args="load pcl/PassThrough /horizontal_object_detector_manager">
      <remap from="~input" to="$(arg INPUT_CLOUD)" />
  </node>

  <node name="ground_estimate" pkg="nodelet" type="nodelet"
        args="load jsk_pcl/OrganizedMultiPlaneSegmentation /horizontal_object_detector_manager">
    <remap from="~input" to="passthrough_filter_for_ground/output" />
  </node>

  <node name="multi_plane_extraction" pkg="nodelet" type="nodelet"
        args="load jsk_pcl/MultiPlaneExtraction /horizontal_object_detector_manager">
    <remap from="~input" to="hsi_filter_red/output" />
    <remap from="~indices" to="ground_estimate/output_refined" />
    <remap from="~input_polygons" to="ground_estimate/output_refined_polygon" />
    <remap from="~input_coefficients" to="ground_estimate/output_refined_coefficients" />
  </node>

  <node name="clustering_red_object" pkg="nodelet" type="nodelet"
        args="load jsk_pcl/EuclideanClustering /horizontal_object_detector_manager">
    <remap from="~input" to="multi_plane_extraction/output" />
   </node>

  <node name="line_detector" pkg="nodelet" type="nodelet"
        args="load jsk_pcl/LineSegmentDetector /horizontal_object_detector_manager">
    <remap from="~input" to="multi_plane_extraction/output" />
    <remap from="~input_indices" to="clustering_red_object/output" />
  </node>

  <node pkg="aerial_robot_perception" type="get_horizontal_object_info.py" name="get_horizontal_object_info" output="screen"/>
</launch>
